test/e2e/redundancy/gateway_failover.go

```diff
@@ -137,118 +136,119 @@ func defaultEndpointType() tcp.EndpointType {
-	// terminated.
-	By(fmt.Sprintf("Setting the gateway label for node %q to false", initialGatewayNode.Name))
-	f.SetGatewayLabelOnNode(framework.ClusterA, initialGatewayNode.Name, false)
-
-	By(fmt.Sprintf("Verifying that the gateway %q was deleted", initialGatewayNode.Name))
-	Eventually(deleteCh, framework.TestContext.OperationTimeout).Should(Receive(Equal(initialGatewayNode.Name)))
-
-	// Ensure the new gateway pod is started before we run the connectivity tests to eliminate possible timing issue where,
-	// after deleting the old pod, we actually run the connectivity test against the old gateway instance before k8s has
-	// a chance to react to stop the process/container etc.
-	var newGatewayPod *v1.Pod
-	for retries := 1; retries < 10; retries++ {
-		newGatewayPod = f.AwaitSubmarinerGatewayPod(framework.ClusterA)
-		if newGatewayPod.Spec.NodeName == initialGatewayNode.Name {
-			time.Sleep(5 * time.Second)
-		} else {
-			break
+	// Fetch the new state of the gateway active pod to ensure the switch to active completed successfully.
+	By("Checking the gateway pod has updated the HA status label as active")
+	var updatedActiveGatewayPod string
+	var updatedActiveGatewayNode string
+
+	updatedGwPods := f.AwaitPodsByAppLabel(framework.ClusterIndex(primaryCluster), framework.SubmarinerGateway, framework.TestContext.SubmarinerNamespace, gwPodCount)
+	for i := range updatedGwPods.Items {
+		if updatedGwPods.Items[i].Labels[gatewayStatusLabel] == gatewayStatusActive {
+			updatedActiveGatewayPod = updatedGwPods.Items[i].Name
+			updatedActiveGatewayNode = updatedGwPods.Items[i].Labels[gatewayNodeLabel]
 		}
 	}
-	Expect(newGatewayPod.Spec.NodeName).To(Equal(initialNonGatewayNode.Name),
-		"The new gateway pod is not running on the expected node")
-	By(fmt.Sprintf("Found new submariner gateway pod %q", newGatewayPod.Name))
-
-	By(fmt.Sprintf("Waiting for the new pod %q to report as active and fully connected", newGatewayPod.Name))
-	f.AwaitGatewayFullyConnected(framework.ClusterA, resource.EnsureValidName(newGatewayPod.Spec.NodeName))
+	Expect(updatedActiveGatewayPod).To(Equal(initialPassiveGatewayPod),
+		"The passive gateway pod has not switched to be active as expected")
+	Expect(updatedActiveGatewayNode).To(Equal(initialPassiveGatewayNode),
+		"The new active gateway pod is not located on expected node")
+	By(fmt.Sprintf("Found new active submariner gateway pod %q on node %q", updatedActiveGatewayPod, updatedActiveGatewayNode))
 
-	By(fmt.Sprintf("Checking that gateway pod %q updated the HA status label", newGatewayPod.Name))
-
-	newGatewayPod = f.AwaitSubmarinerGatewayPod(framework.ClusterA)
-	Expect(newGatewayPod.Labels[gatewayStatusLabel]).To(Equal(gatewayStatusActive))
+	By(fmt.Sprintf("Waiting for the new pod %q to report as active and fully connected", updatedActiveGatewayPod))
+	f.AwaitGatewayFullyConnected(framework.ClusterIndex(primaryCluster), resource.EnsureValidName(updatedActiveGatewayNode))
```

Comment:
We want to ensure a new GW pod on a new node has started but again we don't want to assume it's `initialPassiveGatewayPod`. Also it's cleaner if we can avoid checks for "kind". I think we can do similar as before where we keep querying `AwaitPodsByAppLabel` until we find a pod running on a different node than before. Also we can pass in -1 for `expectedCount` in which case it will ignore the count.

Suggested code to replace L188-221 above:

```
var newGatewayPod *v1.Pod
for retries := 1; retries <= 30; retries++ {
    gwPods := f.AwaitPodsByAppLabel(framework.ClusterIndex(primaryCluster), framework.SubmarinerGateway,
			framework.TestContext.SubmarinerNamespace, -1)
    newGatewayPod = nil
    for i := range gwPods.Items {
        pod := &gwPods.Items[i]
        if pod.Labels[gatewayStatusLabel] == gatewayStatusActive {
            if newGatewayPod != nil {
                framework.Errorf("Found 2 active gateway pods: %q and %q", newGatewayPod.Name, pod.Name)
                newGatewayPod = nil
                break
            }

            if pod.Spec.NodeName != initialGatewayNode.Name {
                newGatewayPod = pod
            }
        }
    }

    if newGatewayPod != nil {
        break
    }

    time.Sleep(2 * time.Second)
}

Expect(newGatewayPod).ToNot(BeNil(), "Did not find a new gateway pod running on a different node")
	
By(fmt.Sprintf("Found new submariner gateway pod %q", newGatewayPod.Name))

By(fmt.Sprintf("Waiting for the new pod %q to report as active and fully connected", newGatewayPod.Name))
f.AwaitGatewayFullyConnected(framework.ClusterA, 
                   resource.EnsureValidName(newGatewayPod.Spec.NodeName))

```

Reply:
Thank you.
Will apply the suggested change.

Reply:
Applied